### DISCUSSION TITLE
Elegoo Mars 4 DLP Exposure Test?

### DISCUSSION THREAD

**Discussion by bryanseger** [sn4k3/UVtools#689] [General]
Recently got a Mars 4 DLP this past week. I see that the UVtools was recently updated to work with the new .goo file format, so that was encouraging to see as we use UVtools across multiple printers to find correct exposure settings. However, using the exposure test on the Mars 4 DLP isn't working as expected. The test will begin, but where I expect a small 'hop' in Z (0.1mm) between each exposure time, it moves the entire Z layer height (currently running at .25mm). Using Voxeldance Tango to do the slicing (which has its own issues, but I don't think that's related here). 

I expect this is an issue on Elegoo's side instead of UVtools (similar to how UVtools doesn't work with Anycubic printers), but, I wanted to reach out on here just in case we're missing something.

**Comment by sn4k3** [comment]
In theory the format structure is ready to support such, in practice I don't know, that was never tested before.
Have you let the test finish? Sometimes Z hope can be higher than defined due firmware.

**Comment by bryanseger** [comment]
Yes, we ran two tests so far with similar results. For example, if I'm doing a 3 piece exposure test (e.g., 1s, 2s, & 3s), at .025mm layers, the 1s piece begins, and before the 2s piece begins, the head moves an entire .025mm upwards and seemingly ignores the .01mm 'hop', and again another .025mm movement to the 3s piece. When it returns to the 1s piece, the build head is now .075mm higher instead of the intended .025mm, resulting in the entire test being much thicker in the Z than expected.

Just to try something quick, I turned off the Z hop option in UVtools but did not get any different results.

**Reply by sn4k3** [reply]
If you want upload the test file so I can check the internal structure of the file.

**Comment by bryanseger** [comment]
Attached is a file sliced for the Mars 4 Max, but I see similar issues here as I did on the Mars 4 DLP (I think the issues are all related to the new .goo format). In the attached case, the 45s wait-before-cure time is ignored by the machine, and also ignores the transition from 45s-->15s wait-before-cure time, and only follows the 15s wait time for every layer (both bottom and normal). 
[seger test.zip](https://github.com/sn4k3/UVtools/files/11334059/seger.test.zip)

Do you see any issue in the attached file?

**Reply by sn4k3** [reply]
I see the problem... `PerLayerSettings` set flag was inverted resulting in file not follow per layer settings when it should and otherwise too. (Also affecting AnyCubic files)
Test this:

[seger_test_fixed.zip](https://github.com/sn4k3/UVtools/files/11334754/seger_test_fixed.zip)

**Reply by bryanseger** [reply]
Right on, thanks for the quick response!

Our 4 Max just started a 7 hour build, so it'll need to wait until tomorrow to test it out.

Was the issue on my end (with how I set it up), or was the issue part of UVtools? I wonder if that explains the issues we were seeing with the Mars 4 DLP as well.

**Reply by sn4k3** [reply]
Problem is on UVtools logic, it was inverted. It affects all .goo and anycubic files.

**Reply by bryanseger** [reply]
Gotcha--would that possibly explain issues we were seeing trying to run a UVtools exposure test on both the 4 Max and 4 DLP? 

Attached is what I should have included in my original message when I started this chain. I just tried this file on the Mars 4 DLP and I see two issues with running it.

First, when it goes from one test to the next (there are 9 test strips total), the Z is moving the full .025mm each time from strip to strip. For example, the first test is at 1s, the second test is at 6s, and the Z is moving a full .025mm between these two. By the time it comes back around to the next 1s test, the Z has moved a full .225mm (9 x .025mm) instead of just .025mm. I am verifying this with a Mitutoyo digital indicator.

Second, the 1s exposure time set in the first test strip is used for all 9 strips, instead of following the time that I had set in UVtools (a 5s increment, just for this test file)--so. all 9 test strips here follow the 1s exposure time, even though UVtools does show the time changes correctly (at least on my end).

I'm hoping this is something you can tweak in UVtools and isn't an issue with the new .GOO format. 
[042623 seger goo test 2.zip](https://github.com/sn4k3/UVtools/files/11335567/042623.seger.goo.test.2.zip)
![4DLP](https://user-images.githubusercontent.com/4185392/234652162-a7420a24-b777-4f0e-9e3f-0ae1e740b2a7.jpg)

(ignore the 0.019mm reading on the indicator...this image is just an example of how I have this set up)

**Reply by sn4k3** [reply]
> Gotcha--would that possibly explain issues we were seeing trying to run a UVtools exposure test on both the 4 Max and 4 DLP?

Yes, `042623 seger goo test 2` also have same problem.
It will be fixed on next release, but confirm if test now works: [042623 seger goo test 2.zip](https://github.com/sn4k3/UVtools/files/11336093/042623.seger.goo.test.2.zip)

**Reply by bryanseger** [reply]
I tested your new test file, and while I did see some change, I'm still seeing issues. 

Let me preface this with the fact that I have seen Z height issues on both the 4 Max and 4 DLP from Tango alone, where the Z height would abruptly shift between layers. I haven't studied it in great detail yet, but it seems that the "slow up" and "fast up" steps are followed, but then the "fast down" is ignored and only "slow down" is followed, so whatever distance is set for the "fast down" setting is ignored in the physical Z height movement. Meaning, several millimeters of difference occurs from one layer to the next. I have see this issue straight from Tango, but I did not have this issue yesterday prior to trying your 'fix test files' (by some workaround that we figured out in Tango). 

On the test setup today, I'm now running a ~2" stroke Mitutoyo gauge so I can better capture the full range of Z motion. To summarize, this UVtools exposure test is set up to run 3 layers of full strips, and then proceed to 9 individual test strips, ranging from 1s to 41s with a 5s step in between. Below is the data from this run. I stopped collecting data after 2 full rounds of the 9 individual strips.

You can see the large Z jump between layers 1-2, 2-3, and 3-4, where layer 4 is the first individual exposure strip. However, instead of starting at 1s, it receives the same 10s 'bottom layer' exposure time. Layer 5, or the second individual exposure strip, starts at 1s, but it's a full .025mm higher in Z instead of being the same as layer 4. The 5s exposure time increment is followed correctly from here on out for a while, but the Z is jumping ~.025mm every time instead of remaining the same.

Layer 13 is the first individual exposure strip but it gets the full 41s of exposure time instead of the 1s it should get, and the Z has jumped another ~3.5mm instead of the expected .025mm. Layer 14 reverts back to a .025mm Z movement and down to 1s, where there after it follows the 5s exposure time increment but jumps ~.025mm every layer.

Hopefully that makes sense! I'm hoping you see a possible fix on your issue, and/or maybe Tango is spitting out some incorrect values.

I am more than happy to run another test for you! Hopefully we can get to the bottom of this and make UVtools work on both the Mars 4 printers.

![mars 4 DLP data](https://user-images.githubusercontent.com/4185392/234987726-6b495ce8-5ed0-4feb-a954-dbd69a3c8b70.jpg)


![mars4 DLP setup](https://user-images.githubusercontent.com/4185392/234987976-f86c5e4d-88c9-4301-85ef-14325bcb3e1a.jpg)

**Reply by sn4k3** [reply]
Everything on file looks correct now. The shift you see on time is strange but remember that UVtools layers are 0 started and then is possible to confuse if printer shows you layer in 1 started and you compare with UVtools.
I will release the patch issuing this problem and you can further test from scratch.

Note that slicers don't use `PerLayerSettings` beside Chitubox Pro if you configure different heights. So out of Tango it should follow global values. You can observe `PerLayerSettings` property on UVtools being `false` on that files

**Reply by bryanseger** [reply]
I should have stated that the layer count is just me giving them a generic number, not necessarily what the printer or UVtools says the layer number is. When I say 'layer 4', that's the first individual strip layer that should start at 1s, and I was visually watching it, so the time shift is definitely an issue on this particular test file.

I did not configure different heights, at least not intentionally. I think there is some odd behavior coming out of Tango that hopefully will be fixed on their end soon as more users start testing the new Mars 4 printers. 


Awesome, looking forward to your patch update and testing this some more. Thanks!

**Reply by sn4k3** [reply]
Maybe Tango is not the problem. If UVtools shows you the correct values then that's not a problem in Slicer. You can confirm slicing with PrusaSlicer and my profiles.
Patch is up.

**Reply by bryanseger** [reply]
I have updated to the patch and ran another test, but I still got similar problems as what I saw yesterday. Attached is the file that I just tested. 

Below is a link to a quick video of this test in action. Hopefully you can see how far the Z movement is out over the first few layers, how the 1s exposure time is ignored for the first individual test strip, and how there's a ~.025mm Z movement between the individual test strips instead of remaining at the same Z height.

Does this reveal anything or help on your end?

https://photos.app.goo.gl/R38G7pgobyAZSTxE7

[042823 UVtools FIX test3.zip](https://github.com/sn4k3/UVtools/files/11356619/042823.UVtools.FIX.test3.zip)

**Reply by sn4k3** [reply]
> Does this reveal anything or help on your end?

No, everything on that file looks correct now. This should be a firmware problem now.

**Reply by renqh123** [reply]
Hi guys, I just saw this issue. I checked the slicing file in the inside, as Tiago said, there are no issues. I will check in my Mars 4 DLP and if there are any similar issues, I will contact the firmware developer for correction. Thank you all.

**Reply by renqh123** [reply]
Friends, I'm back. I have confirmed that it is a firmware issue and a firmware update is required. I have contacted the firmware developer.

**Reply by renqh123** [reply]
@bryanseger @sn4k3

**Reply by renqh123** [reply]
In addition, the exposure time for the fourth layer is not the expected 1s, but a longer time. I have confirmed with the firmware developer that the value of the transition layer in this slicing file affects it, resulting in the first ordinary layer being recognized as the transition layer, so the exposure time is not as expected for 1s.

**Reply by renqh123** [reply]
![image](https://github.com/sn4k3/UVtools/assets/41279743/05315fe7-0b3b-4378-bea4-83230f082abc)
So, Tiago needs to complete a patch and set the transition layer of the exposure test file to 0. 
You can always trust Tiago.

**Reply by sn4k3** [reply]
@renqh123 You using a old UVtools version, that transition layer count is fixed on latter versions. However even if that version shows that number they are not set on file when test was first made. If you reopen the file and then made some save it will be set, otherwise it will be 0 if file was only made/saved once.

**Reply by renqh123** [reply]
> @renqh123 You using a old UVtools version, that transition layer count is fixed on latter versions. However even if that version shows that number they are not set on file when test was first made. If you reopen the file and then made some save it will be set, otherwise it will be 0 if file was only made/saved once.

Tiago, you win, I have confirmed the number of transition layers and it is indeed 0. I have already submitted this issue to the firmware developer.
[doge]

**Reply by bryanseger** [reply]
> > @renqh123 You using a old UVtools version, that transition layer count is fixed on latter versions. However even if that version shows that number they are not set on file when test was first made. If you reopen the file and then made some save it will be set, otherwise it will be 0 if file was only made/saved once.
> 
> Tiago, you win, I have confirmed the number of transition layers and it is indeed 0. I have already submitted this issue to the firmware developer. [doge]

So you reached out to Elegoo about this? Did they confirm there's a firmware issue?

**Reply by renqh123** [reply]
> > > @renqh123 You using a old UVtools version, that transition layer count is fixed on latter versions. However even if that version shows that number they are not set on file when test was first made. If you reopen the file and then made some save it will be set, otherwise it will be 0 if file was only made/saved once.
> > 
> > 
> > Tiago, you win, I have confirmed the number of transition layers and it is indeed 0. I have already submitted this issue to the firmware developer. [doge]
> 
> So you reached out to Elegoo about this? Did they confirm there's a firmware issue?

Yes, I contacted Elegoo and confirmed that it was a firmware issue. At this moment, I have received the updated lift/return issues for the Elego Mars 4 DLP, as well as the test firmware for the fourth layer exposure time issue. Do you need it? I can share it with you.

**Reply by renqh123** [reply]
@bryanseger

**Reply by bryanseger** [reply]
Yes please, can you attach what you had received onto here?

**Reply by renqh123** [reply]
[elegoo_dlp.zip](https://github.com/sn4k3/UVtools/files/11477885/elegoo_dlp.zip)
Import the Bin file into the USB drive, insert the printer, and then restart the printer.
You can see that there is a green update progress bar when starting.
If not, please let me know.

**Reply by bryanseger** [reply]
That worked on my test file! I haven't tried it yet from a UVtools standpoint, but I hope to get to that tomorrow. Elegoo needs to list this updated firmware on their website.

**Reply by bryanseger** [reply]
Ok, I have tested UVtools now on the 4 DLP, and with the updated firmware, everything is working as it should!

UVtools is still not functioning correctly on the 4 Max, but I think a firmware fix from Elegoo will correct it based on what we've seen with the 4 DLP.

**Reply by renqh123** [reply]
You're right, the firmware of DLP is a test version, and the official firmware update is done by ELEGOO.
